File: leetcode/subtree-of-another-tree/subtree-of-another-tree-iter.rb
# LeetCode #572 - Subtree of Another Tree
# https://leetcode.com/problems/subtree-of-another-tree/
# By hashing, linear O(M + N) runtime, implemented iteratively. The state
# machine implements the same logic as in subtree-of-another-tree.rb.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {TreeNode} sub_root
# @return {Boolean}
def is_subtree(root, sub_root)
  table = {} # val, left_id, right_id => id
  traverse(root, table, true)
  traverse(sub_root, table, false)
end

def traverse(root, table, add_if_absent)
  stack = []
  stack << Frame.new(root) if root
  last_id = nil # "return" cell

  until stack.empty?
    frame = stack.last

    case frame.state
    when :go_left
      if frame.node
        frame.state = :go_right
        stack << Frame.new(frame.node.left)
      else
        last_id = 0
        stack.pop
      end

    when :go_right
      frame.left_id = last_id
      frame.state = :retreat
      stack << Frame.new(frame.node.right)

    when :retreat
      key = [frame.node.val, frame.left_id, last_id]
      last_id = table[key]

      unless last_id
        return false unless add_if_absent
        table[key] = last_id = table.size + 1
      end

      stack.pop
    end
  end

  true
end

# A stack frame for iteratively implemented recursive binary tree traversal.
class Frame
  attr_reader :node
  attr_accessor :left_id
  attr_accessor :state

  def initialize(node)
    @node = node
    @left_id = nil
    @state = :go_left
  end
end

File: leetcode/subtree-of-another-tree/subtree-of-another-tree-iter.cpp
// LeetCode #572 - Subtree of Another Tree
// https://leetcode.com/problems/subtree-of-another-tree/
// By hashing, linear O(M + N) runtime, implemented iteratively. The state
// machine implements the same logic as in subtree-of-another-tree-rec.cpp.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    [[nodiscard]] static bool
    isSubtree(const TreeNode* root, const TreeNode* subRoot) noexcept;
};

namespace {
    struct Key {
        int val;
        int left_id;
        int right_id;
    };

    [[nodiscard]]
    constexpr bool operator==(const Key& lhs, const Key& rhs) noexcept
    {
        return lhs.val == rhs.val
                && lhs.left_id == rhs.left_id
                && lhs.right_id == rhs.right_id;
    }

    [[maybe_unused, nodiscard]]
    constexpr bool operator!=(const Key& lhs, const Key& rhs) noexcept
    {
        return !(lhs == rhs);
    }
}

template<>
struct std::hash<Key> {
    [[nodiscard]] constexpr size_t operator()(const Key& key) const noexcept
    {
        auto ret = bias;
        ret = ret * multiplier + key.val;
        ret = ret * multiplier + key.left_id;
        ret = ret * multiplier + key.right_id;
        return ret;
    }

private:
    static constexpr auto bias = size_t{1907};

    static constexpr auto multiplier = size_t{131'071};
};

namespace {
    using Table = unordered_map<Key, int>;

    enum class State { go_left, go_right, retreat };

    struct Frame {
        explicit constexpr Frame(const TreeNode* const node_) noexcept
            : node{node_} { }

        const TreeNode* node;

        int left_id {-1};

        State state {State::go_left};
    };

    template<bool AddIfAbsent>
    bool traverse(conditional_t<AddIfAbsent, Table&, const Table&> table,
                  const TreeNode* const root) noexcept
    {
        auto stack = std::stack<Frame>{};
        if (root) stack.emplace(root);

        auto last_id = -1; // "return" cell

        while (!empty(stack)) {
            auto& frame = stack.top();

            switch (frame.state) {
            case State::go_left:
                if (frame.node) {
                    frame.state = State::go_right;
                    stack.emplace(frame.node->left);
                } else {
                    last_id = 0;
                    stack.pop();
                }

                break;

            case State::go_right:
                frame.left_id = last_id;
                frame.state = State::retreat;
                stack.emplace(frame.node->right);
                break;

            case State::retreat:
                const auto key = Key{frame.node->val, frame.left_id, last_id};

                if (const auto it = table.find(key); it != end(table)) {
                    last_id = it->second;
                } else if constexpr (AddIfAbsent) {
                    last_id = static_cast<int>(size(table)) + 1;
                    table.emplace(key, last_id);
                } else {
                    return false;
                }

                stack.pop();
            }
        }

        return true;
    }
}

bool Solution::isSubtree(const TreeNode* const root,
                         const TreeNode* const subRoot) noexcept
{
    auto table = Table{};
    traverse<true>(table, root);
    return traverse<false>(table, subRoot);
}
